Mail section: Compose Mail, Reply, Reply All, Forward.

zelfroster opened the pull request:
Fixed Compose Mail -
- fixed issues and improved UI
- add recipients in To: input easily and manage.
![1688394194_Jul03_19:53:14](https://github.com/RetroShare/RSNewWebUI/assets/93514460/e3fb8514-3dc1-4c47-a1ae-87a031413a6f)

Removed config appearance, but kept UI fixes from that PR.
Cc and Bcc input remaining in compose mail.

defnax commented:
nice composer looks no more clean :)

zelfroster commented:
- Added Cc and Bcc option in composer.
- Grouped the mail data into a single object.
- Made code reusable for To, Cc and Bcc.

https://github.com/RetroShare/RSNewWebUI/assets/93514460/097f8174-1098-4db2-93a8-302558525be0

I wanted to make it like gmail, where cc and bcc are not shown by default and show when clicked, but will do it later. 
![1688456578_Jul04_13:12:58](https://github.com/RetroShare/RSNewWebUI/assets/93514460/f2323b35-1dfb-44d4-b430-8ab0f843c3f1)

I will work on reply feature first.

defnax commented:
one thing is needed too, attach files in composer

zelfroster commented:
> one thing is needed too, attach files in composer

Alright, I will try to implement it as well.

defnax commented:
ok and picture attachments too into the text field

zelfroster commented:
> ok and picture attachments too into the text field

I looked into it. This will be hard to implement, as It would need to be done differently than the composer already is. Will do it later.

csoler commented:
It's perfectly fine to step away from one particular job for a while in order to let the ideas mature. Don't worry: when you come back to it, you'll probably fix it.

defnax commented:
> 
>  simplescreenrecorder-2023-07-04_13.04.02.mp4 
> I wanted to make it like gmail, where cc and bcc are not shown by default and show when clicked, but will do it later. ![1688456578_Jul04_13:12:58](https://user-images.githubusercontent.com/93514460/250803437-f2323b35-1dfb-44d4-b430-8ab0f843c3f1.png)
> 
> I will work on reply feature first.

Hi, but hte composer needs too close after send automaticly , you closed manualy the composer, that  right?

zelfroster commented:
> > simplescreenrecorder-2023-07-04_13.04.02.mp4
> > I wanted to make it like gmail, where cc and bcc are not shown by default and show when clicked, but will do it later. ![1688456578_Jul04_13:12:58](https://user-images.githubusercontent.com/93514460/250803437-f2323b35-1dfb-44d4-b430-8ab0f843c3f1.png)
> > I will work on reply feature first.
> 
> Hi, but the composer needs to close after send automaticly , you closed manualy the composer, that right?

Can be done 👍🏼

zelfroster commented:
Reply Mail Finally working, along with some fixes.
- Optimized code for toggling Recipients view in Mail view, when large number of recipients.
- Replaced iframe which was used earlier with a contenteditable div so that message is viewed easily and in the future, image insertion directly in mail compose might be possible.
- Fixed order of Mail lists, now mails will not switch order when switching routes and coming back to them.
- And, made the mail popup close automatically using setState like functionality in React.

Here is a preview of Mail Reply.
https://github.com/RetroShare/RSNewWebUI/assets/93514460/250210d8-b5f8-474e-a8f0-3402306163e6

Also, I had a query about How actual mail content is parsed or composed in Qt side?
And, Need suggestions on correctly formatting the reply message conversation.

defnax commented:
look here maybe useful for reply header/ quoted message:
[MessageComposer *MessageComposer::replyMsg(const std::string &msgId, bool all)
](https://github.com/RetroShare/RetroShare/blob/7a3c9f7efe06a119c02d64cb7b2e5404d1700755/retroshare-gui/src/gui/msgs/MessageComposer.cpp#L1200)

https://github.com/RetroShare/RetroShare/blob/7a3c9f7efe06a119c02d64cb7b2e5404d1700755/retroshare-gui/src/gui/msgs/MessageComposer.cpp#L1095

defnax commented:
and we has color highlighter on qt for mail replies 
https://github.com/RetroShare/RetroShare/commit/e4929456128a11e497fe4614092099149aba2b1f

and quouted text with ">" ">>" ...
https://github.com/RetroShare/RetroShare/blob/7a3c9f7efe06a119c02d64cb7b2e5404d1700755/retroshare-gui/src/util/HandleRichText.cpp#L1248

![image](https://github.com/RetroShare/RSNewWebUI/assets/9952056/70815b71-4631-4343-8516-c0a9c700ab1d)

zelfroster commented:
Sorry, I was busy buying PC components and assembling them. @csoler This PR can be reviewed.

rottencandy commented on webui-src/app/main.js:
```diff
@@ -87,6 +87,9 @@ const Layout = () => {
         }),
         m('.tab-content', vnode.children),
       ]),
+      m('#notification-container'),
+      m('#modal-container'),
```
Do we need to move the modal element to inside the main mount? It might interfere with it's vdom since it's using a different render tree.

rottencandy replied:
Yeah that would be better I think. Normally we should try to avoid mixing m.mount & m.render in the same tree

zelfroster replied:
Done 👍🏼

zelfroster commented:
I also wanted to discuss caching in webui for large set of data, I have so far researched and found out two methods which include using browser's `sessionStorage` or `indexedDb` api.

Since same data is fetched multiple times, we can store them for every session so that there are less number of requests for same data.

rottencandy reviewed (commented):
I haven't reviewed the CSS very closely but everything else looks good

rottencandy commented:
Caching sounds good! You could also look into `localStorage` API for more persistence.
